lib/Support/ModelHelpers.cpp

```
@@ -35,278 +33,123 @@
 
-    case 64:
-      ModelType.UnqualifiedType() = Model.getPrimitiveType(Generic, 8);
-      break;
+    revng_assert(IntType->getIntegerBitWidth() % 8 == 0);
+    return model::PrimitiveType::makeGeneric(IntType->getIntegerBitWidth() / 8);
 
-    case 80:
-      ModelType.UnqualifiedType() = Model.getPrimitiveType(Generic, 10);
-      break;
-
-    case 96:
-      ModelType.UnqualifiedType() = Model.getPrimitiveType(Generic, 12);
-      break;
-
-    case 128:
-      ModelType.UnqualifiedType() = Model.getPrimitiveType(Generic, 16);
-      break;
-
-    default:
-      revng_abort("Found an LLVM integer with a size that is not a power of "
-                  "two");
-    }
   } else {
-    revng_abort("Only integer types can be directly converted from LLVM types "
-                "to C types.");
+    revng_abort("Only integer and pointer types can be directly converted from "
+                "LLVM types to C types.");
   }
-
-  return ModelType;
 }
 
-QualifiedType deserializeFromLLVMString(llvm::Value *V,
-                                        const model::Binary &Model) {
+model::UpcastableType deserializeFromLLVMString(llvm::Value *V,
+                                                const model::Binary &Model) {
   // Try to get a string out of the llvm::Value
   llvm::StringRef BaseTypeString = extractFromConstantStringPtr(V);
+  auto ParsedType = deserialize<model::UpcastableType>(BaseTypeString);
+  if (not ParsedType)
+    revng_abort("Could not deserialize the ModelGEP base type");
 
-  // Try to parse the string as a qualified type (aborts on failure)
-  QualifiedType ParsedType;
-  {
-    llvm::yaml::Input YAMLInput(BaseTypeString);
-    YAMLInput >> ParsedType;
-    std::error_code EC = YAMLInput.error();
-    if (EC)
-      revng_abort("Could not deserialize the ModelGEP base type");
-  }
-  ParsedType.UnqualifiedType().setRoot(&Model);
-  revng_assert(ParsedType.UnqualifiedType().isValid());
+  revng_assert(!ParsedType->isEmpty());
```

Please report a good error message for this too.

---

Done: `"Type in a LLVM constant string was set to `model::UpcastableType::empty()`. How did it slip through?"`